internal/dcache/clustermap/clustermap.go

```diff
@@ -33,55 +33,353 @@
 
 package clustermap
 
-import "github.com/Azure/azure-storage-fuse/v2/internal/dcache"
+import (
+        "encoding/json"
+        "fmt"
+        "math"
+        "os"
+        "path/filepath"
+        "strconv"
+        "strings"
 
-type ClusterMap interface {
+        "github.com/Azure/azure-storage-fuse/v2/common"
+        "github.com/Azure/azure-storage-fuse/v2/common/log"
+        "github.com/Azure/azure-storage-fuse/v2/internal/dcache"
+)
 
-	//It will be used to close the notification channel
-	stop()
+func Stop() {
+        clusterMap.stop()
+}
+
+// Update is used by ClusterManager to notify clustermap whenever there's an updated local clustermap.
+func Update() {
+        clusterMap.update()
+}
+
+// It will return online MVs Map <mvName, MV> as per local cache copy of cluster map.
+func GetActiveMVs() map[string]dcache.MirroredVolume {
+        return clusterMap.getActiveMVs()
+}
+
+// It will return the cache config as per local cache copy of cluster map.
+func GetCacheConfig() *dcache.DCacheConfig {
+        return clusterMap.getCacheConfig()
+}
+
+// It will return degraded MVs Map <mvName, MV> as per local cache copy of cluster map.
+func GetDegradedMVs() map[string]dcache.MirroredVolume {
+        return clusterMap.getDegradedMVs()
+}
+
+// It will return all the RVs Map <rvName, RV> for this particular node as per local cache copy of cluster map.
+func GetMyRVs() map[string]dcache.RawVolume {
+        return clusterMap.getMyRVs()
+}
+
+// It will return all the RVs Map <rvName, rvState> for the particular mvName as per local cache copy of cluster map.
+func GetRVs(mvName string) map[string]dcache.StateEnum {
+        return clusterMap.getRVs(mvName)
+}
+
+// It will check if the given nodeId is online as per local cache copy of cluster map.
+func IsOnline(nodeId string) bool {
```

Comment:
Why do we need these wrappers? If dcache code is going to create an object of ClusterMap then they can directly call the member functions instead of calling some globally exposed methods which internally then do nothing but to forward the call to cluster-map object. Is the only objective here is to make this singleton ?

Reply:
Encapsulation is the primary objective here. Components other than clustermap (and clustermanager) do not need to worry about the clustermap, and how to get the latest one, etc. They simply call these functions exported by the clustermap package to find the info they want to know about the cluster. Internally, clustermap package takes care of refreshing the local clustermap object periodically (infact it's the clustermanager that refreshes and then notifies clustermap). So you can see how all those details are hidden from components which will use the clustermap package as a clustermanager client.